libs/platform/user/kernel_um.cpp

```diff
@@ -2,18 +2,147 @@
+     *
+     * @param[in] context The address of a EX_RUNDOWN_REF structure.
+     */
+    void
+    initialize_rundown_ref(_In_ const void* context)
+    {
+        std::unique_lock lock(_lock);
+
+        // Re-initialize the entry if it already exists.
+        if (_rundown_ref_counts.find((uint64_t)context) != _rundown_ref_counts.end()) {
+            _rundown_ref_counts.erase((uint64_t)context);
+        }
+
+        _rundown_ref_counts[(uint64_t)context] = {false, 0};
+    }
+
+    /**
+     * @brief Acquire a rundown ref for the given context.
+     *
+     * @param[in] context The address of a EX_RUNDOWN_REF structure.
+     * @return true Rundown has not started.
+     * @return false Rundown has started.
+     */
+    bool
+    acquire_rundown_ref(_In_ const void* context)
+    {
+        std::unique_lock lock(_lock);
+
+        // Fail if the entry is not initialized.
+        if (_rundown_ref_counts.find((uint64_t)context) == _rundown_ref_counts.end()) {
+            throw std::runtime_error("rundown ref table not initialized");
+        }
+
+        // Check if the entry is already rundown.
+        if (std::get<0>(_rundown_ref_counts[(uint64_t)context])) {
+            return false;
+        }
+
+        // Increment the ref count if the entry is not rundown.
+        std::get<1>(_rundown_ref_counts[(uint64_t)context])++;
+
+        return true;
+    }
+
+    /**
+     * @brief Release a rundown ref for the given context.
+     *
+     * @param[in] context The address of a EX_RUNDOWN_REF structure.
+     */
+    void
+    release_rundown_ref(_In_ const void* context)
+    {
+        std::unique_lock lock(_lock);
+
+        // Fail if the entry is not initialized.
+        if (_rundown_ref_counts.find((uint64_t)context) == _rundown_ref_counts.end()) {
+            throw std::runtime_error("rundown ref table not initialized");
+        }
+
+        if (std::get<1>(_rundown_ref_counts[(uint64_t)context]) == 0) {
+            throw std::runtime_error("rundown ref table already released");
+        }
+
+        std::get<1>(_rundown_ref_counts[(uint64_t)context])--;
+
+        if (std::get<1>(_rundown_ref_counts[(uint64_t)context]) == 0) {
+            _rundown_ref_cv.notify_all();
+        }
+    }
+
+    /**
+     * @brief Wait for the rundown ref count to reach 0 for the given context.
+     *
+     * @param[in] context The address of a EX_RUNDOWN_REF structure.
+     */
+    void
+    wait_for_rundown_ref(_In_ const void* context)
+    {
+        std::unique_lock lock(_lock);
+
+        // Fail if the entry is not initialized.
+        if (_rundown_ref_counts.find((uint64_t)context) == _rundown_ref_counts.end()) {
+            throw std::runtime_error("rundown ref table not initialized");
+        }
+
+        auto& [rundown, ref_count] = _rundown_ref_counts[(uint64_t)context];
+        rundown = true;
+        // Wait for the ref count to reach 0.
+        _rundown_ref_cv.wait(lock, [&ref_count] { return ref_count == 0; });
```

Comment:
Per C++ referrence, ```std::condition_variable::wait``` can be awakened 'spuriously'.  Generally, the solution is for the caller to use the ```cv.wait(...)``` call in a loop to ensure the condition is indeed true.  Am I missing something here?

This is (was?) a pretty common pattern on u**x systems that I've used in the past and was wondering if the windows C++ stdlib and/or STL implementation fixes this internally.

The gory details are @ https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables, with a suggestion to use 'atomics' to get rid of the loop

Reply:
Article says always use a predicate, which this code does. Can you please explain the issue in more detail?

Reply:
Ah.  Turns out the overloaded version of ```std::condition_variable::wait()``` (that takes a predicate) is equivalent to the 'loop' pattern I'd mentioned, i.e. it loops over the predicate return value to determine a _genuine_ 'return-from-wait' state.